geoapify/batch.py

```diff
@@ -0,0 +1,210 @@
+        for i in range(math.ceil(len(inputs) / batch_len)):
+            batches.append(inputs[i * batch_len:(i + 1) * batch_len])
+
+        result_urls = []
+        for i, batch in enumerate(batches):
+            params = {'format': 'json'}
+            if parameters is not None:
+                params = {**params, **parameters}
+            data = {
+                'api': api,
+                'params': params,
+                'inputs': batch
+            }
+            try:
+                response = requests.post(
+                    'https://api.geoapify.com/v1/batch?apiKey={}'.format(self._api_key), json=data,
+                    headers=self._HEADERS)
+            except requests.exceptions.RequestException as e:
+                raise SystemExit(e)
+            if response.status_code not in (200, 202):
+                raise ValueError(f'The service failed to create the job for batch {i}' +
+                                 f' - check input range {i * batch_len}:{min((i + 1) * batch_len, len(inputs))}.')
+            url = response.json()['url']
+            result_urls.append(url)
+            time.sleep(0.1)
+
+        return result_urls
+
+    def monitor_batch_jobs_and_get_results(self, sleep_time: int, result_urls: List[str]) -> List[dict]:
+        """Monitors completion of each batch processing job and returns/stores results.
+
+        Previous POST requests started batch processing jobs on geopify.com servers. Here we monitor the status and
+        return/store results when all jobs succeeded.
+
+        :param sleep_time: time in seconds to sleep between every request for a single job.
+        :param result_urls: list of batch job URLs that are to be monitored.
+        :return: batch job results as a list - one element per location.
+        """
+        self._total_number_jobs = len(result_urls)
+
+        with ThreadPoolExecutor(min(10, len(result_urls))) as executor:
+            results = executor.map(lambda x: self._task(url=x, sleep_time=sleep_time), result_urls)
+
+        result_responses = []
+        for result in results:
+            result_responses += result
+
+        return result_responses
+
+    def _task(self, url: str, sleep_time: int) -> List[dict]:
+        job_id = url.split('&apiKey')[0]
+        while True:
+            response = requests.get(url, headers=self._HEADERS).json()
+            try:
+                _ = response['results']
+                with self._lock:
+                    self._number_completed_jobs += 1
+                    self._logger.info(
+                        f'Job {job_id} done - {self._number_completed_jobs}/{self._total_number_jobs} completed.')
+                break
+            except KeyError:
+                self._logger.info(f'Job {job_id} still pending - waiting another {sleep_time} seconds.')
+                time.sleep(sleep_time)
+        return response['results']
+
+    @staticmethod
+    def _get_sleep_time(batch_len: int) -> int:
```

Comment:
You are very right - I thought about this but kept it simple for now. If you have an idea for some quick solution, let me know. Otherwise let's keep as is for now.

Reply:
Just by playing with the numbers, "number_of_addresses ** 0.4", seems not bad

Reply:
I have pushed a suggestion (accidentally pushed it 3 times)

Reply:
Good. So you believe, not the batch size but the total number of inputs is the more determining factor? Anyway, we should not overengineer here now (later maybe :-)).

Maybe we should rather think about how to automatically pick the batch_len parameter.

Just one thing: in case of the batch place details, it needs to be len(inputs) since place_ids is optional (either place_ids or geocodes). And to make this homogeneous, I will use len(inputs) everywhere as input in the sleep_time function.